### PR TITLE
Start Embedded Boundary development

### PR DESCRIPTION
**Pull request by WeiqunZhang**
Modify the build system for embedded boundary (EB) support.  Currently EB is
a compile time option that must be explicitly enabled.  For GNU Make, one
can enable it with `USE_EB=TRUE`, whereas for CMake, `-DAMReX_EB=ON`.  Later
we could decide to enable EB by default with all regular geometry.

Add a simple geometry initialization function, WarpX::InitEB.  By default,
the geometry is all regular.  A few basic types such as box, cylinder,
plane, sphere, etc. are supported via ParmParse runtime parameters.  See
`amrex/Src/EB/AMReX_EB2.cpp` for more details.  Later, we could build more
complex geometry using constructive solid geometry (CSG).  (
https://en.wikipedia.org/wiki/Constructive_solid_geometry ) See
`amrex/Tutorials/EB/GeometryGeneration` for an example of CSG.  There is
also a STL ( https://en.wikipedia.org/wiki/STL_(file_format) ) approach
under development in AMReX.

Add a new member, m_factory, to WarpX class.  This object can be used to
obtain geometry information such as whether a cell is cut, edge centroids,
etc.  Currently we are not using these factories to build MultiFabs for
field data to embed the geometry information into the data containers.  We
could do that later if it is needed or it makes things more convenient.
Nevertheless, this should be sufficient to start the EB development of the
field solver.

It's not clear to me yet how many ghost cells are needed for the geometry
information.  It's currently one, and can be adjusted.

In the future, when particle and embedded boundary interaction is
considered, we can use `amrex::FillSignedDistance` function to obtain signed
distance function on the nodes.  With that information, one should be able
to determine where and when a particle collides with the embedded boundary.

### CODE REVIEW

**Review by atmyers — APPROVED** [review]

**Review by ax3l — APPROVED** [review]
Added additional user-facing CMake logic and some small cleaning :)

File: Source/WarpX.H [hunk 1]
```diff
@@ -877,6 +877,24 @@ private:
...
     amrex::IntVect numprocs{0};
 
+    //
+    // Embeded Boundary
+    //
+
+    // Factory for field data
+    amrex::Vector<std::unique_ptr<amrex::FabFactory<amrex::FArrayBox> > > m_factory;
+
+    amrex::FabFactory<amrex::FArrayBox> const& Factory (int lev) const noexcept {
```
Comment (ax3l) [inline Source/WarpX.H]: Factory is a very general term in programming for a function that constructs classes.
As a general `WarpX::Factory` method it seems a bit technical/non-descript. Do you have a suggestion how we could make it more descriptive for the domain scientists?

E.g. `fieldFactory` or `make_fields` and `make_EBfields` or so?

**Comment by ax3l** [comment]
I don't understand why the PSATD runtime tests crash now... They run on GNUmake and I thought my include re-ordering should not have caused this. But I guess I am wrong on that? Will trigger the PSATD CI again, maybe it was a hickup.

**Comment by ax3l** [comment]
Restarting CI worked :)

File: Source/WarpX.H (hunk 1 above)
Comment (WeiqunZhang) [reply Source/WarpX.H]: I like. `fieldFactory`.  I will make the change.

**Review by ax3l — APPROVED** [review]
Thank you, all aboard! :ship:

**Review by RevathiJambunathan — APPROVED** [review]

**Pull request merged**